Switch to cross-compilation for `linux_aarch64` and `linux_ppc64le`

bdice opened the pull request:
<!--
Thank you for your pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

<!--
Please add any other relevant info below:
-->

bdice committed:
Switch to cross-compilation for `linux_aarch64` and `linux_ppc64le`

bdice commented:
@conda-forge-admin, please rerender

Commit:
MNT: Re-rendered with conda-build 24.9.0, conda-smithy 3.40.1, and conda-forge-pinning 2024.09.30.02.36.09

bdice committed:
Add stdlib('c') and setuptools.

bdice committed:
Merge branch 'cross-compile' of github.com:bdice/ptxcompiler-feedstock into cross-compile

jakirkham commented:
@conda-forge-admin, please re-render

Commit:
MNT: Re-rendered with conda-build 24.9.0, conda-smithy 3.40.1, and conda-forge-pinning 2024.09.30.02.36.09

jakirkham commented on conda-forge.yml:
```diff
@@ -1,3 +1,6 @@
+build_platform:
+  linux_aarch64: linux_64
+  linux_ppc64le: linux_64
```
Cross-compiled builds are running out of space on CI. This is a known issue (particularly with CUDA 11 builds). The typical solution is to more proactively clean up the CI image beforehand using [`azure`'s `free_disk_space`]( https://conda-forge.org/docs/maintainer/conda_forge_yml/#azure ). Have included this suggestion below

```suggestion
azure:
  free_disk_space: true
build_platform:
  linux_aarch64: linux_64
  linux_ppc64le: linux_64
```

Note: This will also need a re-render to take effect

bdice committed:
Free disk space before running.

Co-authored-by: jakirkham <jakirkham@gmail.com>

bdice commented:
@conda-forge-admin, please rerender

Commit:
MNT: Re-rendered with conda-build 24.9.0, conda-smithy 3.40.1, and conda-forge-pinning 2024.09.30.02.36.09

bdice commented:
Hmm. The CI failures are fairly cryptic. @jakirkham I'm not sure if this package would even support cross-compilation, from looking at its setup.py: https://github.com/rapidsai/ptxcompiler/blob/0.8.1/setup.py

jakirkham commented:
The important bit is getting buried

From [this CI build]( https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1041794&view=logs&j=d1321064-f6c3-56b7-0172-6d994d01c836&t=39022612-2533-571e-7b41-94814e7d38bb&l=4000 ), these lines are relevant

```
  /home/conda/feedstock_root/build_artifacts/ptxcompiler_1727730086202/_build_env/bin/aarch64-conda-linux-gnu-c++ -fvisibility-inlines-hidden -fmessage-length=0 -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O3 -pipe -isystem /home/conda/feedstock_root/build_artifacts/ptxcompiler_1727730086202/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/include -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/ptxcompiler_1727730086202/work=/usr/local/src/conda/ptxcompiler-0.8.1 -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/ptxcompiler_1727730086202/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold=/usr/local/src/conda-prefix -isystem /usr/local/cuda/targets/sbsa-linux/include -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /home/conda/feedstock_root/build_artifacts/ptxcompiler_1727730086202/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/include -isystem /usr/local/cuda/targets/sbsa-linux/include -fPIC -I/home/conda/feedstock_root/build_artifacts/ptxcompiler_1727730086202/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/include -I/usr/local/cuda/include -I/home/conda/feedstock_root/build_artifacts/ptxcompiler_1727730086202/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/include/python3.12 -c ptxcompiler/_ptxcompilerlib.cpp -o build/temp.linux-aarch64-cpython-312/ptxcompiler/_ptxcompilerlib.o -Wall -Werror -std=c++11
  /home/conda/feedstock_root/build_artifacts/ptxcompiler_1727730086202/_build_env/bin/aarch64-conda-linux-gnu-c++ -fvisibility-inlines-hidden -fmessage-length=0 -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O3 -pipe -isystem /home/conda/feedstock_root/build_artifacts/ptxcompiler_1727730086202/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/include -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/ptxcompiler_1727730086202/work=/usr/local/src/conda/ptxcompiler-0.8.1 -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/ptxcompiler_1727730086202/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold=/usr/local/src/conda-prefix -isystem /usr/local/cuda/targets/sbsa-linux/include -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /home/conda/feedstock_root/build_artifacts/ptxcompiler_1727730086202/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/include -isystem /usr/local/cuda/targets/sbsa-linux/include -shared -Wl,--allow-shlib-undefined -Wl,-rpath,/home/conda/feedstock_root/build_artifacts/ptxcompiler_1727730086202/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib -Wl,-rpath-link,/home/conda/feedstock_root/build_artifacts/ptxcompiler_1727730086202/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib -L/home/conda/feedstock_root/build_artifacts/ptxcompiler_1727730086202/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib -Wl,--allow-shlib-undefined -Wl,-rpath,/home/conda/feedstock_root/build_artifacts/ptxcompiler_1727730086202/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib -Wl,-rpath-link,/home/conda/feedstock_root/build_artifacts/ptxcompiler_1727730086202/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib -L/home/conda/feedstock_root/build_artifacts/ptxcompiler_1727730086202/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--allow-shlib-undefined -Wl,-rpath,/home/conda/feedstock_root/build_artifacts/ptxcompiler_1727730086202/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib -Wl,-rpath-link,/home/conda/feedstock_root/build_artifacts/ptxcompiler_1727730086202/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib -L/home/conda/feedstock_root/build_artifacts/ptxcompiler_1727730086202/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib -fvisibility-inlines-hidden -fmessage-length=0 -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O3 -pipe -isystem /home/conda/feedstock_root/build_artifacts/ptxcompiler_1727730086202/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/include -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/ptxcompiler_1727730086202/work=/usr/local/src/conda/ptxcompiler-0.8.1 -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/ptxcompiler_1727730086202/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold=/usr/local/src/conda-prefix -isystem /usr/local/cuda/targets/sbsa-linux/include -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /home/conda/feedstock_root/build_artifacts/ptxcompiler_1727730086202/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/include -isystem /usr/local/cuda/targets/sbsa-linux/include build/temp.linux-aarch64-cpython-312/ptxcompiler/_ptxcompilerlib.o -L/home/conda/feedstock_root/build_artifacts/ptxcompiler_1727730086202/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib -L/usr/local/cuda/lib64 -lnvptxcompiler_static -o build/lib.linux-aarch64-cpython-312/ptxcompiler/_ptxcompilerlib.cpython-312-aarch64-linux-gnu.so
  /home/conda/feedstock_root/build_artifacts/ptxcompiler_1727730086202/_build_env/bin/../lib/gcc/aarch64-conda-linux-gnu/11.4.0/../../../../aarch64-conda-linux-gnu/bin/ld: skipping incompatible /usr/local/cuda/lib64/libnvptxcompiler_static.a when searching for -lnvptxcompiler_static
  /home/conda/feedstock_root/build_artifacts/ptxcompiler_1727730086202/_build_env/bin/../lib/gcc/aarch64-conda-linux-gnu/11.4.0/../../../../aarch64-conda-linux-gnu/bin/ld: cannot find -lnvptxcompiler_static: No such file or directory
```

Namely it is trying to statically link `nvptxcompiler_static` for `x86_64` when building an `aarch64` library, which fails as expected

It need to look in a different place to find the `aarch64` static library

jakirkham commented on recipe/meta.yaml:
```diff
@@ -22,9 +22,11 @@ requirements:
     - {{ compiler("c") }}
     - {{ compiler("cxx") }}
     - {{ compiler("cuda") }}
+    - {{ stdlib("c") }}
```
```suggestion
    - {{ stdlib("c") }}
    - python                              # [build_platform != target_platform]
    - cross-python_{{ target_platform }}  # [build_platform != target_platform]
```

bdice commented:
@jakirkham At this point it seems like we would need a patch to ptxcompiler's build logic, or to try to make the native (not-cross) builds in #37 work. I'm not sure how to diagnose the native build issues. I'm also okay with deferring this until RAPIDS needs to ship Python 3.13.

jakirkham commented on recipe/meta.yaml:
```diff
@@ -11,7 +11,7 @@ source:
+  number: 3
   skip: true  # [(not linux) or cuda_compiler_version != "11.8"]
   ignore_run_exports_from:
     - {{ compiler("cuda") }}
```
This should override [the upstream `CUDA_HOME` logic]( https://github.com/rapidsai/ptxcompiler/blob/ae91f8dad5e5ff7103128eaf1a77b923887334ec/setup.py#L25-L40 ) to pick up the appropriate `targets` directory

```suggestion
    - {{ compiler("cuda") }}
  script_env:
    - CUDA_HOME=/usr/local/cuda/targets/sbsa-linux     # [aarch64]
    - CUDA_HOME=/usr/local/cuda/targets/ppc64le-linux  # [ppc64le]
```

jakirkham reviewed (commented):
Think we may need something like this to correct `ptxcompiler`'s search paths for libraries during cross-compilation

jakirkham committed:
Use `cross-python`

jakirkham committed:
Set `CUDA_HOME` for arch targets

jakirkham commented:
@conda-forge-admin, please re-render

jakirkham commented on recipe/meta.yaml:
```diff
@@ -11,20 +11,27 @@ source:
   skip: true  # [(not linux) or cuda_compiler_version != "11.8"]
   ignore_run_exports_from:
     - {{ compiler("cuda") }}
+  script_env:
+    - CUDA_HOME=/usr/local/cuda/targets/sbsa-linux     # [aarch64]
+    - CUDA_HOME=/usr/local/cuda/targets/ppc64le-linux  # [ppc64le]
   script: {{ PYTHON }} -m pip install . -vv
```
```suggestion
  script:
    - export CUDA_HOME="${CUDA_HOME}/targets/sbsa-linux"     # [aarch64]
    - export CUDA_HOME="${CUDA_HOME}/targets/ppc64le-linux"  # [ppc64le]
    - {{ PYTHON }} -m pip install . -vv
```

jakirkham committed:
Set `CUDA_HOME` in `script`

Let `nvcc` configure `CUDA_HOME` first. Then set to `targets`.

jakirkham committed:
Simplify `CUDA_HOME` setting

Use `target_name`, which is used in other recipes, to simplify the
`CUDA_HOME` setting.

jakirkham commented:
We are still missing `libnvptxcompiler_static.a` in the cross-compilation case. Have submitted PR ( https://github.com/conda-forge/conda-forge-ci-setup-feedstock/pull/356 ) to remedy that

Edit: This was incorrect. `libnvptxcompiler_static.a` is already present. Just `ptxcompiler` was not looking in the right place

jakirkham committed:
Set `CUDA_HOME` when cross-compiling only

Otherwise `nvptxcompiler_static` may not be found.

jakirkham committed:
Apply upstream patch to search `$CUDA_HOME/lib`

jakirkham commented on recipe/meta.yaml:
```diff
@@ -9,22 +12,35 @@ package:
 source:
   url: https://github.com/rapidsai/ptxcompiler/archive/refs/tags/{{ version }}.tar.gz
   sha256: 469644ac7d1d37ab82b7213d3aae90820c3d840aa483967b006554501fe725ac
+  patches:
+    ##########################################################
+    # Add `$CUDA_HOME/lib` to library search path.           #
+    #                                                        #
+    # xref: https://github.com/rapidsai/ptxcompiler/pull/40  #
+    ##########################################################
+    - PR_40.patch
```
Ah! We were only searching `$CUDA_HOME/lib64` and not `$CUDA_HOME/lib`. Submitted upstream PR ( https://github.com/rapidsai/ptxcompiler/pull/40 ) to fix. For now added as a patch here

jakirkham commented:
Looks like this is now passing! 🥳

bdice commented:
Thanks @jakirkham. I looked over your changes and they seem fine to me! I will go ahead and merge this.

jakirkham commented:
Happy to help. Thanks Bradley! 🙏
